RubberduckTests/Grammar/ResolverTests.cs

```diff
@@ -6858,6 +6858,115 @@ End Function
             }
         }
 
+        [Category("Grammar")]
+        [Category("Resolver")]
+        [Test]
+        public void ExpressionInNextHasReference_For()
+        {
+            var moduleCode = $@"
+Private Sub Foo()
+   Dim loopIndex As Long
+   For loopIndex = 0 To 42
+      'DoSomething
+   Next loopIndex
+End Sub
+";
+
+            var vbe = MockVbeBuilder.BuildFromSingleStandardModule(moduleCode, out _);
+
+            using (var state = Resolve(vbe.Object))
+            {
+                var loopIndex = state.DeclarationFinder.UserDeclarations(DeclarationType.Variable).Single(declaration => declaration.IdentifierName.Equals("loopIndex"));
+                var loopIndexReferences = loopIndex.References;
+
+                Assert.AreEqual(2, loopIndexReferences.Count());
+            }
+        }
+
+        [Category("Grammar")]
+        [Category("Resolver")]
+        [Test]
+        public void ExpressionInNextHasReference_DoubleFor()
+        {
+            var moduleCode = $@"
+Private Sub Foo()
+   Dim loopIndex As Long
+   Dim otherLoopIndex As Long
+   For loopIndex = 0 To 42
+      For otherLoopIndex = 0 To 23
+         'DoSomething
+   Next otherLoopIndex, loopIndex
+End Sub
+";
+
+            var vbe = MockVbeBuilder.BuildFromSingleStandardModule(moduleCode, out _);
+
+            using (var state = Resolve(vbe.Object))
+            {
+                var loopIndex = state.DeclarationFinder.UserDeclarations(DeclarationType.Variable).Single(declaration => declaration.IdentifierName.Equals("loopIndex"));
+                var otherLoopIndex = state.DeclarationFinder.UserDeclarations(DeclarationType.Variable).Single(declaration => declaration.IdentifierName.Equals("otherLoopIndex"));
+                var loopIndexReferences = loopIndex.References;
+                var otherLoopIndexReferences = otherLoopIndex.References;
+
+                Assert.AreEqual(2, loopIndexReferences.Count());
+                Assert.AreEqual(2, otherLoopIndexReferences.Count());
+            }
+        }
+
+        [Category("Grammar")]
+        [Category("Resolver")]
+        [Test]
+        public void ExpressionInNextHasReference_ForEach()
+        {
+            var moduleCode = $@"
+Private Sub Foo()
+   Dim coll As Collection
+   Dim element As Long
+   For Each element In coll
+      'DoSomething
+   Next element
+End Sub
+";
+
+            var vbe = MockVbeBuilder.BuildFromSingleStandardModule(moduleCode, out _);
+
+            using (var state = Resolve(vbe.Object))
+            {
+                var element = state.DeclarationFinder.UserDeclarations(DeclarationType.Variable).Single(declaration => declaration.IdentifierName.Equals("element"));
+                var elementReferences = element.References;
+
+                Assert.AreEqual(2, elementReferences.Count());
+            }
+        }
+
+        [Category("Grammar")]
+        [Category("Resolver")]
+        [Test]
+        public void ExpressionInNextHasReference_DoubleForEach()
```

Comment:
Do we need one more test to avoid ambiguity in those cases?

```
For i = 0 to 4
  For j = 0 to 23
    For k = 0 to 42
  Next j, k
Next i
```

as the variant where the final `Next` contains no reference to the `i`.

Reply:
I could add such a test, but I do not see the added benefit. Actually, I also do not see any ambiguity here. 

Moreover, could you please clarify your last sentence? I have a really hard time interpreting it.